File: apps/ios/Sources/Data/PrivacyViewSettings.swift
// Sublemonable — Copyright (C) 2026 Sublemonable contributors
// Licensed under the GNU Affero General Public License v3.0 or later.
// See the LICENSE file in the repository root for full license text.
// SPDX-License-Identifier: AGPL-3.0-only

import Foundation

/// v1.5 privacy-view and platform-warning logic (UI layer only — never touches
/// crypto or the message envelope). Mirrors packages/protocol privacy.ts.
///
/// Privacy view blurs message content behind a frosted lemon overlay, revealed
/// only while actively interacting. On a screenshot, the blurred state is what
/// gets captured — partial mitigation on iOS.
public enum RevealMode: String, Codable, CaseIterable {
    case holdToReveal
    case tapTimed
    case tapToggle
}

public enum Platform: String, Codable {
    case ios, android, browser
}

public let tapTimedDurationsSeconds = [5, 10, 30]
public let tapTimedDefaultSeconds = 10

public struct PrivacyViewSettings: Codable, Equatable {
    /// Applies privacy view to every conversation.
    public var globalEnabled: Bool
    /// Per-conversation overrides, keyed by peer ID.
    public var perConversation: [String: Bool]
    public var revealMode: RevealMode
    public var tapTimedSeconds: Int

    public init(globalEnabled: Bool = false,
                perConversation: [String: Bool] = [:],
                revealMode: RevealMode = .holdToReveal,
                tapTimedSeconds: Int = tapTimedDefaultSeconds) {
        self.globalEnabled = globalEnabled
        self.perConversation = perConversation
        self.revealMode = revealMode
        self.tapTimedSeconds = tapTimedSeconds
    }

    /// True if privacy view is active for a given conversation.
    public func active(for peerID: String) -> Bool {
        perConversation[peerID] ?? globalEnabled
    }
}

/// Warning copy for the platform badge — browser clients can't block screenshots.
public enum PlatformWarning {
    public static let contactOnBrowser = "🌐 Messaging via browser — screenshot protection unavailable"
    public static let ownBrowser = "🌐 You're on browser — your messages can be screenshotted"

    public struct Result: Equatable {
        public let show: Bool
        public let copy: String
    }

    /// Whether to warn, and which copy, given who is on what.
    public static func forPlatforms(own: Platform, contact: Platform?) -> Result {
        if own == .browser { return Result(show: true, copy: ownBrowser) }
        if contact == .browser { return Result(show: true, copy: contactOnBrowser) }
        return Result(show: false, copy: "")
    }
}

File: apps/android/app/src/main/java/com/sublemonable/app/data/PrivacyViewSettings.kt
// Sublemonable — Copyright (C) 2026 Sublemonable contributors
// Licensed under the GNU Affero General Public License v3.0 or later.
// See the LICENSE file in the repository root for full license text.
// SPDX-License-Identifier: AGPL-3.0-only

package com.sublemonable.app.data

/**
 * v1.5 privacy-view and platform-warning logic (UI layer only — never touches
 * crypto or the message envelope). Mirrors packages/protocol privacy.ts.
 *
 * Privacy view blurs message content behind a frosted lemon overlay, revealed
 * only while actively interacting. On Android, FLAG_SECURE still hard-blocks
 * screenshots regardless of privacy-view state.
 */
enum class RevealMode { HOLD_TO_REVEAL, TAP_TIMED, TAP_TOGGLE }

enum class Platform { IOS, ANDROID, BROWSER }

val TAP_TIMED_DURATIONS_SECONDS = listOf(5, 10, 30)
const val TAP_TIMED_DEFAULT_SECONDS = 10

data class PrivacyViewSettings(
    /** Applies privacy view to every conversation. */
    val globalEnabled: Boolean = false,
    /** Per-conversation overrides, keyed by peer ID. */
    val perConversation: Map<String, Boolean> = emptyMap(),
    val revealMode: RevealMode = RevealMode.HOLD_TO_REVEAL,
    val tapTimedSeconds: Int = TAP_TIMED_DEFAULT_SECONDS,
) {
    /** True if privacy view is active for a given conversation. */
    fun activeFor(peerId: String): Boolean = perConversation[peerId] ?: globalEnabled
}

/** Warning copy for the platform badge — browser clients can't block screenshots. */
object PlatformWarning {
    const val CONTACT_ON_BROWSER = "🌐 Messaging via browser — screenshot protection unavailable"
    const val OWN_BROWSER = "🌐 You're on browser — your messages can be screenshotted"

    data class Result(val show: Boolean, val copy: String)

    /** Whether to warn, and which copy, given who is on what. */
    fun forPlatforms(own: Platform, contact: Platform?): Result = when {
        own == Platform.BROWSER -> Result(true, OWN_BROWSER)
        contact == Platform.BROWSER -> Result(true, CONTACT_ON_BROWSER)
        else -> Result(false, "")
    }
}
